File: app/src/api/projects.ts
/**
 * 项目接口。
 *
 * 项目管理接口返回注册表或单个项目摘要；项目内接口都以 `projectCode` 为首个必需参数，
 * 直接映射到 URL 中的项目资源，不读取任何隐式“当前项目”状态。
 */
import { request, withQuery } from './client'
import { projectApiPath } from '@/lib/projectRoute'
import type {
  ArtBible,
  Character,
  ProjectConfig,
  ProjectConfigPatch,
  ProjectDirState,
  ProjectFinalizeIn,
  ProjectList,
  ProjectSummary,
  ScanResult,
} from '@/types/api'

/** `sync` 会先扫一遍默认项目根，把用户手动拷进去的目录认领进来。 */
export function listProjects(sync = false): Promise<ProjectList> {
  return request<ProjectList>(withQuery('/api/projects', { sync: sync || null }))
}

/** 选完目录先问一句：已经归另一个项目的话，得先让用户点头再覆盖。 */
export function inspectDir(dirPath: string): Promise<ProjectDirState> {
  return request<ProjectDirState>(withQuery('/api/projects/dir-state', { dir_path: dirPath }))
}

/**
 * 立项第一步：占下目录并切过去，接下来在立项页跟 Agent 对焦。
 *
 * `overwrite` 只在用户对着确认框点过头之后带：它会抹掉目录里旧项目的 `project.json`、
 * `art-bible.md` 与 `.atelier/` 运行库（素材文件不动）。
 */
export function bootstrapProject(dirPath: string, overwrite = false): Promise<ProjectSummary> {
  return request<ProjectSummary>('/api/projects/bootstrap', {
    method: 'POST',
    body: { dir_path: dirPath, overwrite },
  })
}

/** 立项收口：定下名字与代号，后端顺手铺素材目录、git 规则与 art bible。 */
export function finalizeProject(
  projectCode: string,
  payload: ProjectFinalizeIn,
): Promise<ProjectSummary> {
  return request<ProjectSummary>(projectApiPath(projectCode, 'finalize'), {
    method: 'POST',
    body: payload,
  })
}

/** 挂上一个已有的项目目录：换机器、外置盘、同事拷来的都走这里。 */
export function importProject(dirPath: string): Promise<ProjectSummary> {
  return request<ProjectSummary>('/api/projects/import', {
    method: 'POST',
    body: { dir_path: dirPath },
  })
}

/** 从本机移出，磁盘上的文件一个不动——项目目录是用户的资产。 */
export function forgetProject(code: string): Promise<void> {
  return request<void>(projectApiPath(code), { method: 'DELETE' })
}

/** 按 URL 中的项目代号读取摘要。 */
export function readProject(projectCode: string): Promise<ProjectSummary> {
  return request<ProjectSummary>(projectApiPath(projectCode))
}

export function readConfig(projectCode: string): Promise<ProjectConfig> {
  return request<ProjectConfig>(projectApiPath(projectCode, 'config'))
}

export function updateConfig(
  projectCode: string,
  patch: ProjectConfigPatch,
): Promise<ProjectConfig> {
  return request<ProjectConfig>(projectApiPath(projectCode, 'config'), {
    method: 'PUT',
    body: patch,
  })
}

export function readArtBible(projectCode: string): Promise<ArtBible> {
  return request<ArtBible>(projectApiPath(projectCode, 'art-bible'))
}

/** 整篇覆盖保存：art bible 是视觉真相，编辑器给的就是全文。 */
export function writeArtBible(projectCode: string, content: string): Promise<ArtBible> {
  return request<ArtBible>(projectApiPath(projectCode, 'art-bible'), {
    method: 'PUT',
    body: { content },
  })
}

export function scanProject(projectCode: string): Promise<ScanResult> {
  return request<ScanResult>(projectApiPath(projectCode, 'scan'), { method: 'POST' })
}

export function listCharacters(projectCode: string): Promise<Character[]> {
  return request<Character[]>(projectApiPath(projectCode, 'characters'))
}

/** 指定项目 `characters/` 下的分组目录（含空分组）。分组只是文件夹，后端直接读盘。 */
export function listGroups(projectCode: string): Promise<string[]> {
  return request<string[]>(projectApiPath(projectCode, 'groups'))
}

/** 建一个空分组文件夹，返回建完后的最新分组列表。 */
export function createGroup(projectCode: string, path: string): Promise<string[]> {
  return request<string[]>(projectApiPath(projectCode, 'groups'), {
    method: 'POST',
    body: { path },
  })
}

/**
 * 用表单值拼配置提交体。
 *
 * 后端收到 `style` / `defaults` 是整段替换，而表单只认识自己画出来的那几个字段——直接提交
 * 会把用户手写在 `project.json` 里的额外键抹掉。所以这里以读回来的配置为底再盖表单值：
 * 平台不认识的键原样带回去。
 */
export function buildConfigPatch(
  config: ProjectConfig,
  values: {
    name: string
    review_mode: ProjectConfig['review_mode']
    conversation_audit: boolean
    pose_template?: string
    style: Partial<ProjectConfig['style']>
    defaults: Partial<ProjectConfig['defaults']>
  },
): ProjectConfigPatch {
  return {
    name: values.name,
    review_mode: values.review_mode,
    conversation_audit: values.conversation_audit,
    // 空串表示「不用姿态模板」，后端的 null 与它同义，统一成 null 免得存一个空字符串进 json
    pose_template: values.pose_template?.trim() ? values.pose_template.trim() : null,
    style: { ...config.style, ...values.style },
    defaults: { ...config.defaults, ...values.defaults },
  }
}

File: server/src/atelier/api/projects.py
"""项目管理接口。

新建项目分两步：`POST /bootstrap` 只占下目录与项目库（此后就能开会话对焦），
`POST /{project_code}/finalize` 在用户确认名字与代号后铺骨架、git 规则与 art bible。
导入则是只登记。项目目录可以在磁盘任意位置，接口一律收发绝对路径（前端从系统文件对话框
拿到的就是绝对路径），库里也只存绝对路径，口径只有一种。

配置读写全部直通磁盘上的 `project.json`：它是唯一真相，库里不留副本，因此不存在「表和
文件不一致」这种要对账的状态。
"""

from __future__ import annotations

from pathlib import Path

from fastapi import APIRouter, Query, status
from sqlalchemy.orm import Session

from atelier.api.characters import character_out
from atelier.api.deps import CurrentProject, RuntimeDb
from atelier.api.schemas import (
    ArtBibleIn,
    ArtBibleOut,
    CharacterOut,
    GroupCreateIn,
    ProjectBootstrapIn,
    ProjectConfigOut,
    ProjectConfigPatch,
    ProjectDirStateOut,
    ProjectFinalizeIn,
    ProjectImportIn,
    ProjectListOut,
    ProjectSummaryOut,
    ScanResultOut,
)
from atelier.assets import projects
from atelier.assets.projects import ProjectRef
from atelier.settings import get_settings

router = APIRouter(prefix="/api/projects", tags=["projects"])


def _summary_out(item: projects.ProjectSummary) -> ProjectSummaryOut:
    return ProjectSummaryOut(
        code=item.code,
        name=item.name,
        dir_path=item.dir_path,
        managed=item.managed,
        missing=item.missing,
        stage=item.stage,
    )


def _list_out(session: Session) -> ProjectListOut:
    return ProjectListOut(
        projects=[_summary_out(item) for item in projects.list_projects(session)],
        default_root=str(get_settings().assets_dir),
    )


def _summary_for(session: Session, ref: ProjectRef) -> ProjectSummaryOut:
    for item in projects.list_projects(session):
        if item.code == ref.code:
            return _summary_out(item)
    # 走不到：ref 是刚从注册表解析或登记出来的
    return ProjectSummaryOut(
        code=ref.code,
        name=ref.name,
        dir_path=str(ref.dir),
        managed=False,
        missing=False,
    )


@router.get("", response_model=ProjectListOut)
def list_projects(session: RuntimeDb, sync: bool = False) -> ProjectListOut:
    """本机的项目列表。

    `sync=true` 时先扫一遍默认项目根，把用户手动拷进 `assets/` 的项目目录认领进来；别处
    的项目不扫，得显式导入——平台不去遍历用户的磁盘。
    """
    if sync:
        projects.sync_default_root(session)
    return _list_out(session)


@router.get("/dir-state", response_model=ProjectDirStateOut)
def inspect_dir(dir_path: str = Query(min_length=1)) -> ProjectDirStateOut:
    """选完目录先问一句：这块地是不是已经归另一个项目。

    占着就报出占着的是哪几个文件，好让界面在覆盖之前把代价说清。
    """
    state = projects.inspect_dir(Path(dir_path))
    return ProjectDirStateOut(
        occupied=state.occupied, marks=list(state.marks), is_project=state.is_project
    )


@router.post("/bootstrap", response_model=ProjectSummaryOut, status_code=status.HTTP_201_CREATED)
def bootstrap_project(payload: ProjectBootstrapIn, session: RuntimeDb) -> ProjectSummaryOut:
    """选完目录就创建一个立项中的项目，接下来由前端进入其项目 URL 对焦。

    此时只有 `project.json` 与项目库，名字暂时用目录名、代号是临时的。

    目录已经归另一个项目时报 409；用户对着确认框点了覆盖就带 `overwrite=true` 再来一次。
    """
    ref = projects.bootstrap_project(session, Path(payload.dir_path), overwrite=payload.overwrite)
    return _summary_for(session, ref)


@router.post("/{project_code}/finalize", response_model=ProjectSummaryOut)
def finalize_project(
    payload: ProjectFinalizeIn, ref: CurrentProject, session: RuntimeDb
) -> ProjectSummaryOut:
    """立项收口：定下名字与代号，铺素材目录、`.gitignore`、`.gitattributes` 与 art bible。

    已立项的项目重复调也安全：目录与文件都是「缺了才补」。
    """
    fresh = projects.finalize_project(session, ref, name=payload.name, code=payload.code)
    return _summary_for(session, fresh)


@router.post("/import", response_model=ProjectSummaryOut)
def import_project(payload: ProjectImportIn, session: RuntimeDb) -> ProjectSummaryOut:
    """挂上一个已有的项目目录（换机器、外置盘、同事拷来的都走这里）。"""
    ref = projects.import_project(session, Path(payload.dir_path))
    return _summary_for(session, ref)


@router.delete("/{project_code}", status_code=status.HTTP_204_NO_CONTENT)
def forget_project(project_code: str, session: RuntimeDb) -> None:
    """从本机移出项目，**不动磁盘上的任何文件**。项目目录是用户的资产。"""
    projects.forget(session, project_code)


@router.get("/{project_code}/config", response_model=ProjectConfigOut)
def read_project_config(ref: CurrentProject) -> ProjectConfigOut:
    return ProjectConfigOut.model_validate(projects.read_config(ref.dir).model_dump())


@router.put("/{project_code}/config", response_model=ProjectConfigOut)
def update_project_config(
    payload: ProjectConfigPatch, ref: CurrentProject, session: RuntimeDb
) -> ProjectConfigOut:
    """改项目配置：只覆盖表单交上来的字段，其余（含用户手写的额外键）原样留着。"""
    config = projects.read_config(ref.dir)
    patch = payload.model_dump(exclude_unset=True, exclude_none=True)
    if "style" in patch:
        config.style = projects.ProjectStyle.model_validate(patch.pop("style"))
    if "defaults" in patch:
        config.defaults = projects.ProjectDefaults.model_validate(patch.pop("defaults"))
    for key, value in patch.items():
        setattr(config, key, value)
    projects.write_config(ref.dir, config)

    # 目录名不跟着改名（改目录会让所有已存的相对路径失效），只同步索引里的显示名
    projects.rename_in_registry(session, ref.code, config.name)
    return ProjectConfigOut.model_validate(config.model_dump())


@router.get("/{project_code}/art-bible", response_model=ArtBibleOut)
def read_art_bible(ref: CurrentProject) -> ArtBibleOut:
    content = projects.read_art_bible(ref)
    return ArtBibleOut(
        path=str(projects.art_bible_path(ref)),
        content=content,
        forbidden=projects.forbidden_terms(content),
    )


@router.put("/{project_code}/art-bible", response_model=ArtBibleOut)
def write_art_bible(payload: ArtBibleIn, ref: CurrentProject) -> ArtBibleOut:
    """整篇覆盖保存。art bible 是视觉真相，编辑器给的就是全文，不做行级合并。"""
    projects.write_art_bible(ref, payload.content)
    return ArtBibleOut(
        path=str(projects.art_bible_path(ref)),
        content=payload.content,
        forbidden=projects.forbidden_terms(payload.content),
    )


@router.post("/{project_code}/scan", response_model=ScanResultOut)
def scan_project(ref: CurrentProject) -> ScanResultOut:
    """扫 `characters/` 目录同步进项目库：用户直接拷进来的素材靠这个被认领。"""
    result = projects.scan_characters(ref)
    return ScanResultOut(added=result.added, missing=result.missing, total=result.total)


@router.get("/{project_code}/groups", response_model=list[str])
def list_groups(ref: CurrentProject) -> list[str]:
    """当前项目 `characters/` 下的分组目录（含空分组）。分组只是文件夹，直接读盘。"""
    return projects.list_groups(ref)


@router.post(
    "/{project_code}/groups", response_model=list[str], status_code=status.HTTP_201_CREATED
)
def create_group(payload: GroupCreateIn, ref: CurrentProject) -> list[str]:
    """在当前项目建一个空分组文件夹，建完回最新分组列表。"""
    projects.create_group(ref, payload.path)
    return projects.list_groups(ref)


@router.get("/{project_code}/characters", response_model=list[CharacterOut])
def list_characters(ref: CurrentProject) -> list[CharacterOut]:
    """当前项目的人物素材。换项目就是换库，列表天然隔离。"""
    return [character_out(row) for row in projects.character_rows(ref)]


@router.get("/{project_code}", response_model=ProjectSummaryOut)
def read_project(ref: CurrentProject, session: RuntimeDb) -> ProjectSummaryOut:
    """返回 URL 指定项目的摘要；未登记或目录缺失时依赖层直接返回 404。"""
    return _summary_for(session, ref)
